src/Symfony/Component/Serializer/Serializer.php

```diff
@@ -247,34 +248,48 @@ public function supportsDenormalization(mixed $data, string $type, string $forma
     private function getNormalizer(mixed $data, ?string $format, array $context): ?NormalizerInterface
     {
         $type = \is_object($data) ? $data::class : 'native-'.\gettype($data);
-
+        $minCached = CacheableSupport::SupportAlways;
+        $minUncached = CacheableSupport::SupportNever;
         if (!isset($this->normalizerCache[$format][$type])) {
+            $minCached = CacheableSupport::Support;
+            $minUncached = CacheableSupport::SupportNot;
             $this->normalizerCache[$format][$type] = [];
-
             foreach ($this->normalizers as $k => $normalizer) {
                 if (!$normalizer instanceof NormalizerInterface) {
                     continue;
                 }
 
-                if (!$normalizer instanceof CacheableSupportsMethodInterface || !$normalizer->hasCacheableSupportsMethod()) {
-                    $this->normalizerCache[$format][$type][$k] = false;
-                } elseif ($normalizer->supportsNormalization($data, $format, $context)) {
-                    $this->normalizerCache[$format][$type][$k] = true;
+                $support = $this->supportsNormalizationWrapper($normalizer, $data, $format, $context);
+                if (CacheableSupport::SupportNever === $support) {
+                    continue;
+                }
+
+                $this->normalizerCache[$format][$type][$k] = $support;
+                if (CacheableSupport::SupportAlways === $support) {
                     break;
                 }
             }
         }
 
         foreach ($this->normalizerCache[$format][$type] as $k => $cached) {
             $normalizer = $this->normalizers[$k];
-            if ($cached || $normalizer->supportsNormalization($data, $format, $context)) {
+            if ($cached->value >= $minCached->value || ($cached->value > $minUncached->value && $this->supportsNormalizationWrapper($normalizer, $data, $format, $context)->value > CacheableSupport::SupportNot->value)) {
```

Comment:
Not sure if it work like that, but WDYT about this approach instead?
```php
        $type = \is_object($data) ? $data::class : 'native-'.\gettype($data);

        if (!isset($this->normalizerCache[$format][$type])) {
            $this->normalizerCache[$format][$type] = [];

            foreach ($this->normalizers as $k => $normalizer) {
                if (!$normalizer instanceof NormalizerInterface) {
                    continue;
                }

                $support = $this->supportsNormalizationWrapper($normalizer, $data, $format, $context);
                if (CacheableSupport::SupportNever === $support) {
                    continue;
                }

                $this->normalizerCache[$format][$type][$k] = $support;
                if (CacheableSupport::SupportAlways === $support) {
                    break;
                }
            }
        }

        foreach ($this->normalizerCache[$format][$type] as $k => $cached) {
            $normalizer = $this->normalizers[$k];
            if ($cached->supports() || $this->supportsNormalizationWrapper($normalizer, $data, $format, $context)->supports()) {
                return $normalizer;
            }
        }

        return null;
```
This seems more straightforward to me